src/Features/CSharp/Portable/CodeMapper/CSharpMapCodeService.NodeToMap.cs

```diff
@@ -0,0 +1,423 @@
+        /// <returns>The body syntax of the specified node, if it has any.</returns>
+        public static SyntaxNode? GetBodySyntax(SyntaxNode node)
+        {
+            return node switch
+            {
+                (MethodDeclarationSyntax method) => method.Body,
+                (LocalFunctionStatementSyntax function) => function.Body,
+                (ConstructorDeclarationSyntax constructor) => constructor.Body,
+                (StructDeclarationSyntax st) => st,
+                (RecordDeclarationSyntax rc) => rc,
+                (BaseTypeDeclarationSyntax cl) => cl,
+                _ => node.ChildNodes().FirstOrDefault(cn => cn is BlockSyntax),
+            };
+        }
+
+        /// <summary>
+        /// Gets the open brace token and closing brace token of a specified syntax node.
+        /// </summary>
+        /// <param name="node">The syntax node to retrieve the brace tokens from.</param>
+        /// <returns>
+        /// An object of type BraceTokens which contain the open brace token and closing brace token of a specified syntax node.
+        /// Or null when the node doesn't have an open brace and close brace tokens.
+        /// </returns>
+        public static BraceTokens? GetOpenCloseBraceTokens(SyntaxNode node)
+        {
+            var bodySyntax = GetBodySyntax(node);
+            if (bodySyntax is null)
+            {
+                return null;
+            }
+
+            if (bodySyntax is BlockSyntax bs)
+            {
+                return new(bs.OpenBraceToken, bs.CloseBraceToken);
+            }
+
+            if (bodySyntax is StructDeclarationSyntax sds)
+            {
+                return new(sds.OpenBraceToken, sds.CloseBraceToken);
+            }
+
+            if (bodySyntax is RecordDeclarationSyntax rds)
+            {
+                return new(rds.OpenBraceToken, rds.CloseBraceToken);
+            }
+
+            if (bodySyntax is BaseTypeDeclarationSyntax btds)
+            {
+                return new(btds.OpenBraceToken, btds.CloseBraceToken);
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// A list of short circuit types.
+        /// These are excluded types that we know we don't want to compare against
+        /// all other scoped or simple types.
+        /// </summary>
+        private static bool IsExcluded(SyntaxKind kind) => kind is SyntaxKind.CompilationUnit;
+
+        /// <summary>
+        /// The list of supported Scoped nodes.
+        /// </summary>
+        private static bool IsScoped(SyntaxKind kind, out Scope scopeKind)
+        {
+            switch (kind)
+            {
+                case SyntaxKind.ClassDeclaration:
+                case SyntaxKind.InterfaceDeclaration:
+                case SyntaxKind.EnumDeclaration:
+                case SyntaxKind.StructDeclaration:
+                case SyntaxKind.RecordDeclaration:
+                    scopeKind = Scope.Class;
```

Comment:
i don't understnad this.  how are interfaces/enum/structs a class?  Also, this seems to be missing delegates.